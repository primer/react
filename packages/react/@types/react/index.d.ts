// eslint-disable-next-line no-restricted-imports
export * from 'react'

declare module 'react' {
  interface CSSProperties {
    [key: `--${string}`]: string | number | undefined
  }
}
