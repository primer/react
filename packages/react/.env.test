VITEST_SEGFAULT_RETRY=0
VITEST_SKIP_BROWSER=1