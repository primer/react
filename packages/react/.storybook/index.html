<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Primer Preview Page</title>
  <link href="https://unpkg.com/@primer/css@^20.2.4/dist/primer.css" rel="stylesheet" />
</head>
<body class="p-3">
  <div class="Header">
    <h1>Primer Preview Page</h1>
  </div>
  <div class="p-2 m-2 d-flex flex-column">
    <ul class="filter-list">
      <li><a class="filter-item" href="./storybook" class="btn-link mb-2">Storybook</a></li>
      <li><a class="filter-item" href="./vrt-no-flag" class="btn-link mb-2">Visual Regression Testing</a></li>
      <li><a class="filter-item" href="./vrt-all-flags" class="btn-link mb-2">Visual Regression Testing (All Flags
          Enabled)</a></li>
      <li><a class="filter-item" href="./aat-no-flag" class="btn-link mb-2">Automated Accessibility Testing</a></li>
      <li><a class="filter-item" href="./aat-all-flags" class="btn-link mb-2">Automated Accessibility Testing (All Flags
          Enabled)</a></li>
    </ul>
  </div>
</body>
</html>