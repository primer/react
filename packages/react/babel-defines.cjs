const shared = {
  __DEV__: "process.env.NODE_ENV !== 'production'",
}

module.exports = {
  development: shared,
  test: shared,
  production: {
    ...shared,
    __DEV__: 'false',
    'process.env.NODE_ENV': "'production'",
  },
}
