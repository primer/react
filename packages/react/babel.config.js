const defines = require('./babel-defines')
const {isSupported} = require('./script/react-compiler.mjs')

function replacementPlugin(env) {
  return ['babel-plugin-transform-replace-expressions', {replace: defines[env]}]
}

const sharedPlugins = [
  [
    'babel-plugin-react-compiler',
    {
      target: '18',
      sources: isSupported,
    },
  ],
  'macros',
  'dev-expression',
  'add-react-displayname',
  'babel-plugin-styled-components',
  '@babel/plugin-proposal-nullish-coalescing-operator',
  '@babel/plugin-proposal-optional-chaining',
]

function makePresets(moduleValue) {
  return [
    '@babel/preset-typescript',
    [
      '@babel/preset-react',
      {
        modules: moduleValue,
        runtime: 'automatic',
      },
    ],
  ]
}

module.exports = {
  env: {
    development: {
      presets: makePresets(process.env.BABEL_MODULE || false),
      plugins: [...sharedPlugins, replacementPlugin('development')],
    },
    production: {
      presets: makePresets(false),
      plugins: [...sharedPlugins, replacementPlugin('production')],
    },
    test: {
      presets: makePresets('commonjs'),
      plugins: [...sharedPlugins, ['@babel/plugin-transform-modules-commonjs'], replacementPlugin('test')],
    },
  },
}
