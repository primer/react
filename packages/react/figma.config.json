{
  "codeConnect": {
    "include": ["src/**/*.figma.tsx", "src/**/*.tsx"],
    "exclude": ["test/**", "docs/**", "build/**"],
    "react": {
      "importPaths": {
        "./src/*": "@primer/react"
      },
      "paths": {}
    }
  }
}
