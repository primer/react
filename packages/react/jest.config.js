/* eslint-disable github/unescaped-html-literal */

'use strict'

const path = require('node:path')

const ROOT_DIR = path.resolve(__dirname, '..', '..')

/**
 * @type {import('jest').Config}
 */
module.exports = {
  testEnvironment: 'jsdom',
  collectCoverageFrom: ['src/**/*.{js,jsx,ts,tsx}', '!src/stories/**', '!**/*.stories.{js,jsx,ts,tsx}'],
  moduleNameMapper: {
    '\\.css$': 'jest-css-modules',
    // We need to specify this package subpath because it does not provide a `require` conditional export path
    '@oddbird/popover-polyfill/fn': path.join(
      // Note: we use ROOT_DIR here since this dependency is hoisted
      ROOT_DIR,
      'node_modules',
      '@oddbird',
      'popover-polyfill',
      'dist',
      'popover-fn.js',
    ),
  },
  setupFiles: ['<rootDir>/src/utils/test-helpers.tsx'],
  setupFilesAfterEnv: ['<rootDir>/src/utils/test-matchers.tsx', '<rootDir>/src/utils/test-deprecations.tsx'],
  testMatch: ['<rootDir>/**/*.test.[jt]s?(x)', '!**/*.types.test.[jt]s?(x)'],
  transformIgnorePatterns: ['node_modules/(?!@github/[a-z-]+-element|@lit-labs/react|@oddbird/popover-polyfill)'],
}
