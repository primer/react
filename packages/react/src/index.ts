'use client'

export {default as theme} from './theme'
export {get as themeGet} from './constants'
export {default as BaseStyles} from './BaseStyles'
export type {BaseStylesProps} from './BaseStyles'
export {default as ThemeProvider, useTheme, useColorSchemeVar} from './ThemeProvider'
export type {ThemeProviderProps} from './ThemeProvider'

// Layout
export {default as Box} from './Box'
export type {BoxProps} from './Box'
export * from './Button'
export {PageLayout} from './PageLayout'
export type {
  PageLayoutProps,
  PageLayoutHeaderProps,
  PageLayoutContentProps,
  PageLayoutPaneProps,
  PageLayoutFooterProps,
} from './PageLayout'
export {SplitPageLayout} from './SplitPageLayout'
export type {
  SplitPageLayoutProps,
  SplitPageLayoutHeaderProps,
  SplitPageLayoutContentProps,
  SplitPageLayoutPaneProps,
  SplitPageLayoutFooterProps,
} from './SplitPageLayout'

// Hooks
export {default as useDetails} from './hooks/useDetails'
export {default as useSafeTimeout} from './hooks/useSafeTimeout'
export {useOnOutsideClick} from './hooks/useOnOutsideClick'
export type {TouchOrMouseEvent} from './hooks/useOnOutsideClick'
export {useOpenAndCloseFocus} from './hooks/useOpenAndCloseFocus'
export {useOnEscapePress} from './hooks/useOnEscapePress'
export {useOverlay} from './hooks/useOverlay'
export {useConfirm} from './ConfirmationDialog/ConfirmationDialog'
export {useFocusTrap} from './hooks/useFocusTrap'
export type {FocusTrapHookSettings} from './hooks/useFocusTrap'
export {useFocusZone} from './hooks/useFocusZone'
export type {FocusZoneHookSettings} from './hooks/useFocusZone'
export {useRefObjectAsForwardedRef} from './hooks/useRefObjectAsForwardedRef'
export {useResizeObserver} from './hooks/useResizeObserver'
export {useResponsiveValue} from './hooks/useResponsiveValue'

// Utils
export {createComponent} from './utils/create-component'

// Components
export {default as Radio} from './Radio'
export type {RadioProps} from './Radio'
export {ActionList} from './ActionList'
export type {
  ActionListProps,
  ActionListGroupProps,
  ActionListItemProps,
  ActionListLinkItemProps,
  ActionListDividerProps,
  ActionListDescriptionProps,
  ActionListLeadingVisualProps,
  ActionListTrailingVisualProps,
} from './ActionList'
export {ActionMenu} from './ActionMenu'
export type {ActionMenuProps, ActionMenuAnchorProps, ActionMenuButtonProps} from './ActionMenu'
export {AnchoredOverlay} from './AnchoredOverlay'
export type {AnchoredOverlayProps} from './AnchoredOverlay'
export {default as Autocomplete} from './Autocomplete'
export type {AutocompleteMenuProps, AutocompleteInputProps, AutocompleteOverlayProps} from './Autocomplete'
export {default as Avatar} from './Avatar'
export type {AvatarProps} from './Avatar'
export {default as AvatarPair} from './AvatarPair'
export type {AvatarPairProps} from './AvatarPair'
export {default as AvatarStack} from './AvatarStack'
export type {AvatarStackProps} from './AvatarStack'
export {default as BranchName} from './BranchName'
export type {BranchNameProps} from './BranchName'
export {default as Breadcrumbs, Breadcrumb} from './Breadcrumbs'
export type {BreadcrumbsProps, BreadcrumbsItemProps, BreadcrumbProps, BreadcrumbItemProps} from './Breadcrumbs'
export {default as ButtonGroup} from './ButtonGroup'
export type {ButtonGroupProps} from './ButtonGroup'
export {default as Caret} from './Caret'
export type {CircleBadgeProps, CircleBadgeIconProps} from './CircleBadge'
export {default as CircleOcticon} from './CircleOcticon'
export type {CircleOcticonProps} from './CircleOcticon'
export {default as CheckboxGroup} from './CheckboxGroup'
export type {CaretProps} from './Caret'
export {default as CircleBadge} from './CircleBadge'
export {default as CounterLabel} from './CounterLabel'
export type {CounterLabelProps} from './CounterLabel'
export {default as Details} from './Details'
export type {DetailsProps} from './Details'
export {default as Dialog} from './Dialog'
export type {DialogProps, DialogHeaderProps} from './Dialog'
export type {ConfirmationDialogProps} from './ConfirmationDialog/ConfirmationDialog'
export {ConfirmationDialog} from './ConfirmationDialog/ConfirmationDialog'
export {default as Flash} from './Flash'
export type {FlashProps} from './Flash'
export {default as FormControl} from './FormControl'
export {useFormControlForwardedProps} from './FormControl'
export {default as Header} from './Header'
export type {HeaderProps, HeaderItemProps, HeaderLinkProps} from './Header'
export {default as Heading} from './Heading'
export type {HeadingProps} from './Heading'
export {default as Label} from './Label'
export type {LabelProps} from './Label'
export {default as LabelGroup} from './LabelGroup'
export type {LabelGroupProps} from './LabelGroup'
export {default as Link} from './Link'
export type {LinkProps} from './Link'
export {NavList} from './NavList'
export type {
  NavListProps,
  NavListItemProps,
  NavListSubNavProps,
  NavListGroupProps,
  NavListLeadingVisualProps,
  NavListTrailingVisualProps,
  NavListDividerProps,
} from './NavList'
export {default as Octicon} from './Octicon'
export type {OcticonProps} from './Octicon'
export {default as Overlay} from './Overlay'
export type {OverlayProps} from './Overlay'
export {default as Pagehead} from './Pagehead'
export type {PageheadProps} from './Pagehead'
export {default as Pagination} from './Pagination'
export type {PaginationProps} from './Pagination'
export {default as PointerBox} from './PointerBox'
export type {PointerBoxProps} from './PointerBox'
export {default as Popover} from './Popover'
export type {PopoverProps, PopoverContentProps} from './Popover'
export {default as Portal, registerPortalRoot} from './Portal'
export type {PortalProps} from './Portal'
export {ProgressBar} from './ProgressBar'
export type {ProgressBarProps} from './ProgressBar'
export {default as RadioGroup} from './RadioGroup'
export type {RelativeTimeProps} from './RelativeTime'
export {default as RelativeTime} from './RelativeTime'
export {SegmentedControl} from './SegmentedControl'
export {default as Select} from './Select'
export type {SelectProps} from './Select'
export {SelectPanel} from './SelectPanel'
export type {SelectPanelProps} from './SelectPanel'
export {default as SideNav} from './SideNav'
export type {SideNavProps, SideNavLinkProps} from './SideNav'
export {default as Spinner} from './Spinner'
export type {SpinnerProps} from './Spinner'
export {default as StateLabel} from './StateLabel'
export type {StateLabelProps} from './StateLabel'
export {default as SubNav} from './SubNav'
export type {SubNavProps, SubNavLinkProps, SubNavLinksProps} from './SubNav'
export {default as ToggleSwitch} from './ToggleSwitch'
export {default as TabNav} from './TabNav'
export type {TabNavProps, TabNavLinkProps} from './TabNav'
export {default as TextInput} from './TextInput'
export type {TextInputProps} from './TextInput'
export {default as TextInputWithTokens} from './TextInputWithTokens'
export type {TextInputWithTokensProps} from './TextInputWithTokens'
export {default as Text} from './Text'
export type {TextProps} from './Text'
export {default as Timeline} from './Timeline'
export type {
  TimelineProps,
  TimelineBadgeProps,
  TimelineBodyProps,
  TimelineBreakProps,
  TimelineItemsProps,
} from './Timeline'
export {default as Token, IssueLabelToken, AvatarToken} from './Token'
export type {TokenProps} from './Token'
export {default as Tooltip} from './Tooltip/Tooltip'
export type {TooltipProps} from './Tooltip/Tooltip'
export {default as Truncate} from './Truncate'
export type {TruncateProps} from './Truncate'

export {default as Checkbox} from './Checkbox'
export type {CheckboxProps} from './Checkbox'

export {default as Textarea} from './Textarea'
export type {TextareaProps} from './Textarea'

export {TreeView} from './TreeView'
export type {
  TreeViewProps,
  TreeViewItemProps,
  TreeViewSubTreeProps,
  TreeViewVisualProps,
  TreeViewErrorDialogProps,
} from './TreeView'

export {UnderlineNav} from './UnderlineNav'
export type {UnderlineNavProps, UnderlineNavItemProps} from './UnderlineNav'

export {ActionBar} from './ActionBar'
export type {ActionBarProps} from './ActionBar'

// eslint-disable-next-line no-restricted-imports
export {SSRProvider, useSSRSafeId} from './utils/ssr'
export {default as sx, merge} from './sx'
export type {SxProp} from './sx'
