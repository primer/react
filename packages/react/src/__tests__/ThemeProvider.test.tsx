import {render, screen, waitFor} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import 'jest-styled-components'
import React from 'react'
import {Text, ThemeProvider, useColorSchemeVar, useTheme} from '..'

// window.matchMedia() is not implemented by JSDOM so we have to create a mock:
// https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
})

const exampleTheme = {
  colors: {
    text: '#f00',
  },
  colorSchemes: {
    light: {
      colors: {
        text: 'black',
      },
    },
    dark: {
      colors: {
        text: 'white',
      },
    },
    dark_dimmed: {
      colors: {
        text: 'gray',
      },
    },
  },
}

it('respects theme prop', () => {
  const theme = {
    colors: {
      text: '#f00',
    },
    space: ['0', '0.25rem'],
  }

  render(
    <ThemeProvider theme={theme}>
      <Text color="text" mb={1}>
        Hello
      </Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#f00')
  expect(screen.getByText('Hello')).toHaveStyleRule('margin-bottom', '0.25rem')
})

it('has default theme', () => {
  render(
    <ThemeProvider>
      <Text color="fg.default" mb={1}>
        Hello
      </Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toMatchSnapshot()
})

it('inherits theme from parent', () => {
  render(
    <ThemeProvider theme={exampleTheme}>
      <ThemeProvider>
        <Text color="text">Hello</Text>
      </ThemeProvider>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'black')
})

it('defaults to light color scheme', () => {
  render(
    <ThemeProvider theme={exampleTheme}>
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'black')
})

it('defaults to dark color scheme in night mode', () => {
  render(
    <ThemeProvider theme={exampleTheme} colorMode="night">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'white')
})

it('defaults to first color scheme when passed an invalid color scheme name', () => {
  const spy = jest.spyOn(console, 'error').mockImplementationOnce(() => {})

  render(
    <ThemeProvider theme={exampleTheme} dayScheme="foo">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(spy).toHaveBeenCalledWith('`foo` scheme not defined in `theme.colorSchemes`')
  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'black')

  spy.mockRestore()
})

it('respects nightScheme prop', () => {
  render(
    <ThemeProvider theme={exampleTheme} colorMode="night" nightScheme="dark_dimmed">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'gray')
})

it('respects nightScheme prop with colorMode="dark"', () => {
  render(
    <ThemeProvider theme={exampleTheme} colorMode="dark" nightScheme="dark_dimmed">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'gray')
})

it('respects dayScheme prop', () => {
  render(
    <ThemeProvider theme={exampleTheme} colorMode="day" dayScheme="dark" nightScheme="dark_dimmed">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'white')
})

it('respects dayScheme prop with colorMode="light"', () => {
  render(
    <ThemeProvider theme={exampleTheme} colorMode="light" dayScheme="dark" nightScheme="dark_dimmed">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'white')
})

it('works in auto mode', () => {
  render(
    <ThemeProvider theme={exampleTheme} colorMode="auto">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'black')
})

it('works in auto mode (dark)', () => {
  const matchMediaSpy = jest.spyOn(window, 'matchMedia').mockImplementation(query => ({
    matches: true, // enable dark mode
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  }))

  render(
    <ThemeProvider theme={exampleTheme} colorMode="auto">
      <Text color="text">Hello</Text>
    </ThemeProvider>,
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', 'white')

  matchMediaSpy.mockRestore()
})

it('updates when colorMode prop changes', async () => {
  const user = userEvent.setup()

  function App() {
    const [colorMode, setColorMode] = React.useState<'day' | 'night'>('day')
    return (
      <ThemeProvider theme={exampleTheme} colorMode={colorMode}>
        <Text color="text">{colorMode}</Text>
        <button type="button" onClick={() => setColorMode(colorMode === 'day' ? 'night' : 'day')}>
          Toggle
        </button>
      </ThemeProvider>
    )
  }

  render(<App />)

  // starts in day mode (light scheme)
  expect(screen.getByText('day')).toHaveStyleRule('color', 'black')

  await user.click(screen.getByRole('button'))

  await waitFor(() =>
    // clicking the toggle button enables night mode (dark scheme)
    expect(screen.getByText('night')).toHaveStyleRule('color', 'white'),
  )
})

it('updates when dayScheme prop changes', async () => {
  const user = userEvent.setup()

  function App() {
    const [dayScheme, setDayScheme] = React.useState('light')
    return (
      <ThemeProvider theme={exampleTheme} dayScheme={dayScheme}>
        <Text color="text">{dayScheme}</Text>
        <button type="button" onClick={() => setDayScheme(dayScheme === 'light' ? 'dark_dimmed' : 'light')}>
          Toggle
        </button>
      </ThemeProvider>
    )
  }

  render(<App />)

  // starts in day mode (light scheme)
  expect(screen.getByText('light')).toHaveStyleRule('color', 'black')

  await user.click(screen.getByRole('button'))

  await waitFor(() =>
    // clicking the toggle sets the day scheme to dark_dimmed
    expect(screen.getByText('dark_dimmed')).toHaveStyleRule('color', 'gray'),
  )
})

it('updates when nightScheme prop changes', async () => {
  const user = userEvent.setup()

  function App() {
    const [nightScheme, setNightScheme] = React.useState('dark')
    return (
      <ThemeProvider theme={exampleTheme} colorMode="night" nightScheme={nightScheme}>
        <Text color="text">{nightScheme}</Text>
        <button type="button" onClick={() => setNightScheme(nightScheme === 'dark' ? 'dark_dimmed' : 'dark')}>
          Toggle
        </button>
      </ThemeProvider>
    )
  }

  render(<App />)

  // starts in night mode (dark scheme)
  expect(screen.getByText('dark')).toHaveStyleRule('color', 'white')

  await user.click(screen.getByRole('button'))

  await waitFor(() =>
    // clicking the toggle button sets the night scheme to dark_dimmed
    expect(screen.getByText('dark_dimmed')).toHaveStyleRule('color', 'gray'),
  )
})

it('inherits colorMode from parent', async () => {
  const user = userEvent.setup()

  function App() {
    const [colorMode, setcolorMode] = React.useState<'day' | 'night'>('day')
    return (
      <ThemeProvider theme={exampleTheme} colorMode={colorMode}>
        <button type="button" onClick={() => setcolorMode(colorMode === 'day' ? 'night' : 'day')}>
          Toggle
        </button>
        <ThemeProvider>
          <Text color="text">{colorMode}</Text>
        </ThemeProvider>
      </ThemeProvider>
    )
  }

  render(<App />)

  expect(screen.getByText('day')).toHaveStyleRule('color', 'black')

  await user.click(screen.getByRole('button'))

  await waitFor(() => expect(screen.getByText('night')).toHaveStyleRule('color', 'white'))
})

it('inherits dayScheme from parent', async () => {
  const user = userEvent.setup()

  function App() {
    const [dayScheme, setDayScheme] = React.useState('light')
    return (
      <ThemeProvider theme={exampleTheme} colorMode="night" dayScheme={dayScheme}>
        <button type="button" onClick={() => setDayScheme(dayScheme === 'light' ? 'dark_dimmed' : 'light')}>
          Toggle
        </button>
        <ThemeProvider colorMode="day">
          <Text color="text">{dayScheme}</Text>
        </ThemeProvider>
      </ThemeProvider>
    )
  }

  render(<App />)

  expect(screen.getByText('light')).toHaveStyleRule('color', 'black')

  await user.click(screen.getByRole('button'))

  await waitFor(() => expect(screen.getByText('dark_dimmed')).toHaveStyleRule('color', 'gray'))
})

it('inherits nightScheme from parent', async () => {
  const user = userEvent.setup()

  function App() {
    const [nightScheme, setNightScheme] = React.useState('dark')
    return (
      <ThemeProvider theme={exampleTheme} colorMode="day" nightScheme={nightScheme}>
        <button type="button" onClick={() => setNightScheme(nightScheme === 'dark' ? 'dark_dimmed' : 'dark')}>
          Toggle
        </button>
        <ThemeProvider colorMode="night">
          <Text color="text">{nightScheme}</Text>
        </ThemeProvider>
      </ThemeProvider>
    )
  }

  render(<App />)

  expect(screen.getByText('dark')).toHaveStyleRule('color', 'white')

  await user.click(screen.getByRole('button'))

  await waitFor(() => expect(screen.getByText('dark_dimmed')).toHaveStyleRule('color', 'gray'))
})

describe('setColorMode', () => {
  it('changes the color mode', async () => {
    const user = userEvent.setup()

    function ToggleMode() {
      const {colorMode, setColorMode} = useTheme()
      return (
        <button type="button" onClick={() => setColorMode(colorMode === 'day' ? 'night' : 'day')}>
          Toggle
        </button>
      )
    }

    render(
      <ThemeProvider theme={exampleTheme} colorMode="day">
        <Text color="text">Hello</Text>
        <ToggleMode />
      </ThemeProvider>,
    )

    // starts in day mode (light scheme)
    expect(screen.getByText('Hello')).toHaveStyleRule('color', 'black')

    await user.click(screen.getByRole('button'))

    // clicking the toggle button enables night mode (dark scheme)
    expect(screen.getByText('Hello')).toHaveStyleRule('color', 'white')
  })
})

describe('setDayScheme', () => {
  it('changes the day scheme', async () => {
    const user = userEvent.setup()

    function ToggleDayScheme() {
      const {dayScheme, setDayScheme} = useTheme()
      return (
        <button type="button" onClick={() => setDayScheme(dayScheme === 'light' ? 'dark' : 'light')}>
          Toggle
        </button>
      )
    }

    render(
      <ThemeProvider theme={exampleTheme} colorMode="day">
        <Text color="text">Hello</Text>
        <ToggleDayScheme />
      </ThemeProvider>,
    )

    // starts in day mode (light scheme)
    expect(screen.getByText('Hello')).toHaveStyleRule('color', 'black')

    await user.click(screen.getByRole('button'))

    // clicking the toggle button sets day scheme to dark
    expect(screen.getByText('Hello')).toHaveStyleRule('color', 'white')
  })
})

describe('setNightScheme', () => {
  it('changes the night scheme', async () => {
    const user = userEvent.setup()

    function ToggleNightScheme() {
      const {nightScheme, setNightScheme} = useTheme()
      return (
        <button type="button" onClick={() => setNightScheme(nightScheme === 'dark' ? 'dark_dimmed' : 'dark')}>
          Toggle
        </button>
      )
    }

    render(
      <ThemeProvider theme={exampleTheme} colorMode="night">
        <Text color="text">Hello</Text>
        <ToggleNightScheme />
      </ThemeProvider>,
    )

    // starts in night mode (dark scheme)
    expect(screen.getByText('Hello')).toHaveStyleRule('color', 'white')

    await user.click(screen.getByRole('button'))

    // clicking the toggle button sets night scheme to dark_dimmed
    expect(screen.getByText('Hello')).toHaveStyleRule('color', 'gray')
  })
})

describe('useColorSchemeVar', () => {
  it('updates value when scheme changes', async () => {
    const user = userEvent.setup()

    function ToggleMode() {
      const {colorMode, setColorMode} = useTheme()
      return (
        <button type="button" onClick={() => setColorMode(colorMode === 'day' ? 'night' : 'day')}>
          Toggle
        </button>
      )
    }

    function CustomBg() {
      const customBg = useColorSchemeVar(
        {
          light: 'red',
          dark: 'blue',
          dark_dimmed: 'green',
        },
        'inherit',
      )

      return <Text bg={customBg}>Hello</Text>
    }

    render(
      <ThemeProvider theme={exampleTheme} nightScheme="dark_dimmed">
        <CustomBg />
        <ToggleMode />
      </ThemeProvider>,
    )

    expect(screen.getByText('Hello')).toHaveStyleRule('background-color', 'red')

    await user.click(screen.getByRole('button'))

    expect(screen.getByText('Hello')).toHaveStyleRule('background-color', 'green')
  })

  it('supports fallback value', async () => {
    const user = userEvent.setup()

    function ToggleMode() {
      const {colorMode, setColorMode} = useTheme()
      return (
        <button type="button" onClick={() => setColorMode(colorMode === 'day' ? 'night' : 'day')}>
          Toggle
        </button>
      )
    }

    function CustomBg() {
      const customBg = useColorSchemeVar({dark: 'blue'}, 'red')

      return <Text bg={customBg}>Hello</Text>
    }

    render(
      <ThemeProvider theme={exampleTheme}>
        <CustomBg />
        <ToggleMode />
      </ThemeProvider>,
    )

    expect(screen.getByText('Hello')).toHaveStyleRule('background-color', 'red')

    await user.click(screen.getByRole('button'))

    expect(screen.getByText('Hello')).toHaveStyleRule('background-color', 'blue')
  })
})

describe('useTheme().resolvedColorScheme', () => {
  it('is undefined when not in a theme', () => {
    const Component = () => {
      const {resolvedColorScheme} = useTheme()

      return <Text data-testid="text">{resolvedColorScheme}</Text>
    }

    render(<Component />)

    expect(screen.getByTestId('text').textContent).toEqual('')
  })

  it('is undefined when the theme has no colorScheme object', () => {
    const Component = () => {
      const {resolvedColorScheme} = useTheme()

      return <Text data-testid="text">{resolvedColorScheme}</Text>
    }

    render(
      <ThemeProvider theme={{color: 'red'}}>
        <Component />
      </ThemeProvider>,
    )

    expect(screen.getByTestId('text').textContent).toEqual('')
  })

  it('is the same as the applied colorScheme, when that colorScheme is in the theme', () => {
    const Component = () => {
      const {resolvedColorScheme} = useTheme()

      return <Text data-testid="text">{resolvedColorScheme}</Text>
    }

    const schemeToApply = 'dark'

    render(
      <ThemeProvider theme={exampleTheme} colorMode="day" dayScheme={schemeToApply}>
        <Component />
      </ThemeProvider>,
    )

    expect(exampleTheme.colorSchemes).toHaveProperty(schemeToApply)
    expect(screen.getByTestId('text').textContent).toEqual(schemeToApply)
  })

  it('is the value of the fallback colorScheme applied when attempting to apply an invalid colorScheme', () => {
    const spy = jest.spyOn(console, 'error').mockImplementationOnce(() => {})
    const Component = () => {
      const {resolvedColorScheme} = useTheme()

      return <Text data-testid="text">{resolvedColorScheme}</Text>
    }

    const schemeToApply = 'totally-invalid-colorscheme'
    render(
      <ThemeProvider theme={exampleTheme} colorMode="day" dayScheme={schemeToApply}>
        <Component />
      </ThemeProvider>,
    )

    const defaultThemeColorScheme = Object.keys(exampleTheme.colorSchemes)[0]

    expect(spy).toHaveBeenCalledWith('`totally-invalid-colorscheme` scheme not defined in `theme.colorSchemes`')
    expect(defaultThemeColorScheme).not.toEqual(schemeToApply)
    expect(exampleTheme.colorSchemes).not.toHaveProperty(schemeToApply)
    expect(screen.getByTestId('text').textContent).toEqual('light')

    spy.mockRestore()
  })

  describe('nested theme', () => {
    it('is the same as the applied colorScheme, when that colorScheme is in the theme', () => {
      const Component = () => {
        const {resolvedColorScheme} = useTheme()

        return <Text data-testid="text">{resolvedColorScheme}</Text>
      }

      const schemeToApply = 'dark'

      render(
        <ThemeProvider theme={exampleTheme} colorMode="day" dayScheme={schemeToApply}>
          <ThemeProvider>
            <Component />
          </ThemeProvider>
        </ThemeProvider>,
      )

      expect(exampleTheme.colorSchemes).toHaveProperty(schemeToApply)
      expect(screen.getByTestId('text').textContent).toEqual(schemeToApply)
    })

    it('is the value of the fallback colorScheme applied when attempting to apply an invalid colorScheme', () => {
      const spy = jest.spyOn(console, 'error').mockImplementation(() => {})

      const Component = () => {
        const {resolvedColorScheme} = useTheme()

        return <Text data-testid="text">{resolvedColorScheme}</Text>
      }

      const schemeToApply = 'totally-invalid-colorscheme'
      render(
        <ThemeProvider theme={exampleTheme} colorMode="day" dayScheme={schemeToApply}>
          <ThemeProvider>
            <Component />
          </ThemeProvider>
        </ThemeProvider>,
      )

      const defaultThemeColorScheme = Object.keys(exampleTheme.colorSchemes)[0]

      expect(spy).toHaveBeenCalledWith('`totally-invalid-colorscheme` scheme not defined in `theme.colorSchemes`')
      expect(defaultThemeColorScheme).not.toEqual(schemeToApply)
      expect(exampleTheme.colorSchemes).not.toHaveProperty(schemeToApply)
      expect(screen.getByTestId('text').textContent).toEqual('light')

      spy.mockRestore()
    })
  })
})
