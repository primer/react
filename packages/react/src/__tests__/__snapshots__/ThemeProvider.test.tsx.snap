// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has default theme 1`] = `
.c0 {
  color: var(--fgColor-default,var(--color-fg-default,#1F2328));
  margin-bottom: 4px;
}

.c0:where([data-size='small']) {
  font-size: var(--text-body-size-small,0.75rem);
  line-height: var(--text-body-lineHeight-small,1.6666);
}

.c0:where([data-size='medium']) {
  font-size: var(--text-body-size-medium,0.875rem);
  line-height: var(--text-body-lineHeight-medium,1.4285);
}

.c0:where([data-size='large']) {
  font-size: var(--text-body-size-large,1rem);
  line-height: var(--text-body-lineHeight-large,1.5);
}

.c0:where([data-weight='light']) {
  font-weight: var(--base-text-weight-light,300);
}

.c0:where([data-weight='normal']) {
  font-weight: var(--base-text-weight-normal,400);
}

.c0:where([data-weight='medium']) {
  font-weight: var(--base-text-weight-medium,500);
}

.c0:where([data-weight='semibold']) {
  font-weight: var(--base-text-weight-semibold,600);
}

<span
  class="c0"
  color="fg.default"
>
  Hello
</span>
`;
